pandas/_libs/lib.pyx

```diff
@@ -2526,3 +2527,21 @@ def fast_multiget(dict mapping, ndarray keys, default=np.nan):
             output[i] = default
 
     return maybe_convert_objects(output)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def from_nested_dict(object data) -> dict:
+    cdef:
+        object new_data = collections.defaultdict(dict)
+        object index, column, value, dict_iterator
+        dict data_dct, nested_dict
+
+    data_dct = dict(data)
+
+    for index, dict_iterator in data_dct.items():
+        nested_dict = dict(dict_iterator)
```

Comment:
If dict_iterator is a Series, this conversion will be slow. Probably just accept the user's choice IMO and accept a slowdown if dict_iterator is not a dict...